bindings/rust/s2n-tls/src/callbacks.rs

```diff
@@ -118,30 +124,128 @@ pub(crate) fn trigger_async_callback<T: 'static + AsyncCallback>(
-    fn poll_client_hello(&self, connection: &mut Connection) -> Poll<Result<(), Error>>;
+    /// The application can return a `None` to resolve the client_hello_callback
+    /// synchronously or return a `Some(AsyncClientHelloFuture)` if it wants to
+    /// run some asynchronous task beforing resolving the callback.
+    ///
+    /// [`ConfigResolver`], which implements [`AsyncClientHelloFuture`] can be
+    /// returned if the application wants to set a new [`Config`] on the connection.
+    ///
+    /// If the server_name is used to configure the connection then the application
+    /// should call [`connection::Connection::server_name_extension_used()`].
+    fn on_client_hello(
+        // this should NEVER be a `&mut` since its would be possible to
+        // mutate the Config on the Connection. Since one Config can be
+        // used for multiple Connections that would be undefined-behavior
+        &self,
+        connection: &mut Connection,
+    ) -> Option<Pin<Box<dyn AsyncClientHelloFuture>>>;
 }
 
 pub(crate) struct AsyncClientHelloCallback {}
 impl AsyncCallback for AsyncClientHelloCallback {
-    fn poll(&mut self, conn: &mut Connection) -> Poll<Result<(), Error>> {
-        let result = conn
+    fn poll(&self, conn: &mut Connection) -> Poll<Result<(), Error>> {
+        // if there is already a future set on the connection then take it and poll it
+        if let Some(mut fut) = conn.take_client_hello_future() {
+            let waker = conn.waker().ok_or(Error::CALLBACK_EXECUTION)?.clone();
+            let mut ctx = core::task::Context::from_waker(&waker);
+            match fut.as_mut().poll_client_hello(conn, &mut ctx) {
+                Poll::Ready(result) => {
+                    let result = result.and_then(|_| conn.mark_client_hello_cb_done());
+                    return Poll::Ready(result);
+                }
+                Poll::Pending => {
+                    // replace the client_hello_future if it hasn't completed yet
+                    conn.set_client_hello_future(fut);
+                    return Poll::Pending;
+                }
+            }
+        }
+
+        // otherwise call the on_client_hello to make progress. storing the future if
+        // the application provided one.
+        let async_future = conn
             .config()
-            .as_ref()
-            .and_then(|config| config.context().client_hello_callback.as_ref())
-            .map(|callback| callback.poll_client_hello(conn))
-            .unwrap_or(READY_OK);
-        if result == READY_OK {
-            conn.mark_client_hello_cb_done()?;
+            .as_mut()
+            .and_then(|config| config.context_mut().client_hello_callback.as_mut())
+            .and_then(|callback| callback.on_client_hello(conn));
+        match async_future {
+            Some(fut) => {
+                // Store the future on connection. This is Asynchronous resolution.
+                conn.set_client_hello_future(fut);
```

Comment:
we need to poll the future here so it registers the waker with the runtime

Reply:
thats subtle. so this was working prob because some other waker was firing and we registered the waker in the following run?